File: src/cppwasm/KernelEntry.cpp
/**
 * Implementation of individual kernel entry for sparse kernel matrix
 * @file KernelEntry.cpp
 *
 * Represents one contributing FFT bin to a frequency bin in the Constant-Q transform.
 */

#include <complex>
#include <cmath>
#include <string>
#include <stdio.h>
#include "KernelEntry.hpp"

using namespace std;

namespace constantq
{
    complex<double> KernelEntry::multiplier() const { return _multiplier; }
    size_t KernelEntry::fftIndex() const { return _fftIndex; }

    KernelEntry::KernelEntry(size_t fftIndex, complex<double> multiplier)
    {
        _fftIndex = fftIndex;
        _multiplier = multiplier;
    }

    /**
     * Generates string representation for debugging
     * @return Human-readable string with index and multiplier values
     */
    string KernelEntry::toString()
    {
        ostringstream stringStream;
        stringStream << "{ Index: " << _fftIndex << ", Multiplier: "
                     << _multiplier.real() << " + " << _multiplier.imag() << "i }";

        return stringStream.str();
    }
};

File: src/app/model/kernelentry.ts
import Complex from './complex';

/**
 * represents an entry in the sparse kernel:
 * the fft index to multiply by and the complex multiplier
 */

export class KernelEntry {
  // the index within the fft to utilize
  readonly fftIndex: number;
  // the multiplier to use
  readonly multiplier: Complex;

  /**
   * creates a new KernelEntry item
   * @param fftIndex      the index within the fft to utilize
   * @param multiplier    the multiplier to apply to this index
   */
  constructor(fftIndex: number, multiplier: Complex) {
    this.fftIndex = fftIndex;
    this.multiplier = multiplier;
  }

  /**
   * the string representation of this KernelEntry
   */
  public toString() {
    return '{ Index: ' + this.fftIndex + ', Multiplier: ' + this.multiplier + ' }';
  }
}
